src/nbl/video/CVulkanLogicalDevice.h

```diff
@@ -542,74 +530,105 @@ class CVulkanLogicalDevice final : public ILogicalDevice
-
-                        VkImageView vk_imageView = static_cast<const CVulkanImageView*>(pDescriptorWrites[i].info[j].desc.get())->getInternalObject();
-
-                        vk_imageInfos[imageInfoOffset + j].sampler = vk_sampler;
-                        vk_imageInfos[imageInfoOffset + j].imageView = vk_imageView;
-                        vk_imageInfos[imageInfoOffset + j].imageLayout = static_cast<VkImageLayout>(descriptorWriteImageInfo.imageLayout);
-                    }
-                    else
-                    {
-                        vk_imageInfos[imageInfoOffset + j] = dummyInfo;
-                    }
-                }
-
-                vk_writeDescriptorSets[i].pImageInfo = vk_imageInfos.data() + imageInfoOffset;
-                imageInfoOffset += pDescriptorWrites[i].count;
+                    vk_imageInfos.push_back(dummyInfo);
             } break;
 
             case asset::IDescriptor::EC_BUFFER_VIEW:
             {
                 VkBufferView dummyBufferView = static_cast<const CVulkanBufferView*>(pDescriptorWrites[i].info[0].desc.get())->getInternalObject();
 
+                vk_writeDescriptorSets[i].pTexelBufferView = reinterpret_cast<VkBufferView*>(vk_bufferViews.size());
+                for (uint32_t j = 0u; j < pDescriptorWrites[i].count; ++j)
+                    vk_bufferViews.push_back(dummyBufferView);
+            } break;
+
+            case asset::IDescriptor::EC_ACCELERATION_STRUCTURE:
+            {
+                auto& writeAS = vk_writeDescriptorSetAS[i];
+                writeAS = { VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET_ACCELERATION_STRUCTURE_KHR, nullptr };
+                writeAS.accelerationStructureCount = pDescriptorWrites[i].count;
+
+                vk_writeDescriptorSets[i].pNext = &writeAS;
+
+                writeAS.pAccelerationStructures = reinterpret_cast<VkAccelerationStructureKHR*>(vk_accelerationStructures.size());
                 for (uint32_t j = 0u; j < pDescriptorWrites[i].count; ++j)
+                    vk_accelerationStructures.push_back({});
+            } break;
+
+            default:
+                assert(!"Invalid code path.");
+            }
+        }
+
+        for (uint32_t i = 0u; i < descriptorWriteCount; ++i)
+        {
+            switch (pDescriptorWrites[i].info->desc->getTypeCategory())
+            {
+            case asset::IDescriptor::E_CATEGORY::EC_BUFFER:
+            {
+                vk_writeDescriptorSets[i].pBufferInfo = reinterpret_cast<size_t>(vk_writeDescriptorSets[i].pBufferInfo) + vk_bufferInfos.data();
+
+                const auto* infoSrc = pDescriptorWrites[i].info;
+                auto* infoDst = const_cast<VkDescriptorBufferInfo*>(vk_writeDescriptorSets[i].pBufferInfo);
+                for (uint32_t j = 0; j < pDescriptorWrites[i].count; ++j, ++infoSrc, ++infoDst)
                 {
-                    if (pDescriptorWrites[i].info[j].info.buffer.size)
+                    if (infoSrc->info.buffer.size)
```

Comment:
why do you check this?

Reply:
Looks like I was handling for the case when we want to update only a select number of array elements in `pDescriptorWrites[i].count` elements (those who have a valid `info` struct associated with them) which I think is not allowed in Vulkan.

Anyway this comes from the [previous version](https://github.com/Devsh-Graphics-Programming/Nabla/blob/b2a17a706dc2ed29fc39fbc6e5861aa0efc94abd/src/nbl/video/CVulkanLogicalDevice.h#L519).
Maybe I misjudged when writing that, we don't have any use case for this i.e. can I always assume that all `pDescriptorWrites[i].count` `info` structs will be valid?

Reply:
> Maybe I misjudged when writing that, we don't have any use case for this i.e. can I always assume that all pDescriptorWrites[i].count info structs will be valid?

Yes, otherwise you're binding the very first item to all subsequent bindings in a binding array.

(You'd be assuming the first one is valid anyway)

Reply:
> Yes, otherwise you're binding the very first item to all subsequent bindings in a binding array.

Yes, the idea was to update them with garbage because they were not dynamically used.

> (You'd be assuming the first one is valid anyway)

Hence the `assert`s [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/d65ffd5c505a4aec3ed1a5581f05e4c0b3580a7b/src/nbl/video/CVulkanLogicalDevice.h#L490). :)

But now all of this can go since we have established that all `pDescriptorWrites[i].count` will be valid (assuming they pass validation).

Reply:
Resolved.